File: playwright-test-demo/pages/login_page.py
"""
LoginPage -- Page Object for the Herokuapp login form.

Target: https://the-internet.herokuapp.com/login
Valid credentials: tomsmith / SuperSecretPassword!
"""

import logging

from pages.base_page import BasePage
from playwright.sync_api import Page

logger = logging.getLogger(__name__)


class LoginPage(BasePage):
    """Page object encapsulating selectors and actions for /login."""

    # -- Selectors --------------------------------------------------------- #
    USERNAME_INPUT: str = "#username"
    PASSWORD_INPUT: str = "#password"
    LOGIN_BUTTON: str = "button[type='submit']"
    FLASH_MESSAGE: str = "#flash"
    LOGOUT_BUTTON: str = "a[href='/logout']"
    PAGE_HEADING: str = "h2"

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    # -- Actions ----------------------------------------------------------- #

    def open(self) -> "LoginPage":
        """Navigate to the login page."""
        self.navigate("/login")
        return self

    def enter_username(self, username: str) -> "LoginPage":
        """Fill in the username field.

        Args:
            username: Value to type.
        """
        self.fill(self.USERNAME_INPUT, username)
        return self

    def enter_password(self, password: str) -> "LoginPage":
        """Fill in the password field.

        Args:
            password: Value to type.
        """
        self.fill(self.PASSWORD_INPUT, password)
        return self

    def click_login(self) -> "LoginPage":
        """Click the Login button."""
        self.click(self.LOGIN_BUTTON)
        return self

    def login(self, username: str, password: str) -> "LoginPage":
        """Perform a complete login.

        Args:
            username: Username to enter.
            password: Password to enter.
        """
        logger.info("Logging in as '%s'", username)
        self.enter_username(username)
        self.enter_password(password)
        self.click_login()
        return self

    # -- State readers ----------------------------------------------------- #

    def get_flash_message(self) -> str:
        """Return the text of the flash notification banner."""
        self.wait_for_element(self.FLASH_MESSAGE)
        return self.get_text(self.FLASH_MESSAGE).strip()

    def is_logged_in(self) -> bool:
        """Return ``True`` if the logout button is visible."""
        return self.is_element_visible(self.LOGOUT_BUTTON)

    def get_page_heading(self) -> str:
        """Return the main ``<h2>`` heading text."""
        return self.get_text(self.PAGE_HEADING)

    def click_logout(self) -> "LoginPage":
        """Click the Logout button."""
        logger.info("Logging out")
        self.click(self.LOGOUT_BUTTON)
        return self

File: selenium-java-demo/src/test/java/pages/LoginPage.java
package pages;

import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Page object for the Herokuapp login page.
 * <p>
 * Target: <a href="https://the-internet.herokuapp.com/login">/login</a>
 */
public class LoginPage extends BasePage {

    // -- Selectors -------------------------------------------------------- //
    private static final By USERNAME_FIELD = By.id("username");
    private static final By PASSWORD_FIELD = By.id("password");
    private static final By LOGIN_BUTTON   = By.cssSelector("button[type='submit']");
    private static final By FLASH_MESSAGE  = By.id("flash");
    private static final By LOGOUT_BUTTON  = By.cssSelector("a[href='/logout']");
    private static final By PAGE_HEADING   = By.tagName("h2");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    // -- Actions ---------------------------------------------------------- //

    @Step("Open login page")
    public LoginPage open() {
        navigateTo("/login");
        return this;
    }

    @Step("Enter username: {username}")
    public LoginPage enterUsername(String username) {
        type(USERNAME_FIELD, username);
        return this;
    }

    @Step("Enter password")
    public LoginPage enterPassword(String password) {
        type(PASSWORD_FIELD, password);
        return this;
    }

    @Step("Click Login button")
    public LoginPage clickLogin() {
        click(LOGIN_BUTTON);
        return this;
    }

    @Step("Login as '{username}'")
    public LoginPage login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
        // Wait for the form submission to settle. The flash banner is present
        // on both success and failure, so this works for positive + negative
        // paths without forcing callers to add their own wait.
        waitForVisible(FLASH_MESSAGE);
        return this;
    }

    @Step("Click Logout button")
    public LoginPage clickLogout() {
        click(LOGOUT_BUTTON);
        return this;
    }

    // -- State readers ---------------------------------------------------- //

    public String getFlashMessage() {
        return getText(FLASH_MESSAGE).trim();
    }

    public boolean isLoggedIn() {
        return isDisplayed(LOGOUT_BUTTON);
    }

    public String getPageHeading() {
        return getText(PAGE_HEADING).trim();
    }

    public boolean isUsernameFieldDisplayed() {
        return isDisplayed(USERNAME_FIELD);
    }

    public boolean isPasswordFieldDisplayed() {
        return isDisplayed(PASSWORD_FIELD);
    }

    public boolean isLoginButtonDisplayed() {
        return isDisplayed(LOGIN_BUTTON);
    }
}
